docs: Add module development guide for west (#112656)
zephyrproject-rtos/zephyr: doc/develop/west/module-development.rst

Before Running ``west update``
==============================

Before running ``west update``, it is good practice to:

* **Commit or stash all local changes**. While ``west update`` will not
  overwrite committed work on local branches, having a clean working tree
  prevents confusion.
* **Check the current state** with ``git status`` and ``west status`` to
  understand which projects have local modifications.

Example:

.. code-block:: shell

   # Check status in the current module
   git status

   # Check status in all projects
   west status

   # Run west update
   west update

   # If west left your branch behind, check it out again
   git checkout my-feature

   # After update, check if your branch needs to be rebased
   git log --oneline HEAD..manifest-rev

.. _west-update-parameters:

``west update`` Parameters for Module Development
*************************************************

For details about options such as ``--keep-descendants`` and ``--rebase``,
see :ref:`west-update` in the :ref:`west-built-in-cmds` documentation.

.. _west-module-dev-scenarios:

Common Development Scenarios
****************************

.. _west-module-dev-scenario-simple-fix:

> Scenario 1: Simple Bug Fix in a Module

Reviewer: This has a significant overlap with "Step-by-Step Guide". Having both does not add value.

======================================

You want to fix a bug in a Zephyr module (e.g., ``modules/lib/zcbor``)
and submit the fix upstream.

1. Navigate to the module:

   .. code-block:: shell

      cd modules/lib/zcbor

2. Update your workspace before starting work:

   .. code-block:: shell

      west update

3. Create a branch for your fix:

   .. code-block:: shell

      git checkout -b fix-uart-timeout

   .. note::

>       If you are submitting the fix upstream to the module's own repository,

Reviewer: Gut feeling: "upstream" says it all - no need to say "to the module's own repository":

```suggestion
      If you are submitting the fix upstream,
```

      consider basing your branch on the module's latest upstream branch
      (e.g., ``main``) rather than on the revision pinned by the Zephyr
      manifest. Fetch it from the module's upstream remote if necessary.

4. Make your changes and commit:

   .. code-block:: shell

      git add .
      git commit -s

5. Push to your fork and create a pull request:

   .. code-block:: shell

      git remote add my-fork https://github.com/your-username/zcbor.git
      git push my-fork fix-uart-timeout

.. _west-module-dev-scenario-downstream:

> Scenario 2: Maintaining a Downstream Fork

Reviewer: a) This section assumes that the user already has a forked downstream module. How to do this is already documented in e.g. https://docs.zephyrproject.org/latest/develop/west/manifest.html#west-manifest-ex1-3.
b) This section does not show anything other than a regular Git workflow.

For this reason, I do not see how this section adds value.

=========================================

You maintain a downstream fork of a module and want to periodically sync with
upstream.

1. Add your downstream remote and update the workspace:

   .. code-block:: shell

      cd modules/lib/zcbor
      git remote add downstream https://github.com/your-company/zcbor.git
      git fetch downstream

>       west update

Reviewer: When a module is already a downstream fork, then there is no point in adding another "downstream" remote. If anything at all, one would add the upstream remote.

      git checkout downstream-v1.0
      git merge manifest-rev   # or: git rebase manifest-rev

4. Push your downstream branch:

   .. code-block:: shell

      git push downstream downstream-v1.0

.. _west-module-dev-scenario-multiple-modules:

Scenario 3: Developing Across Multiple Modules
==============================================

You are working on a feature that spans multiple modules (e.g., a new API
in Zephyr core and its implementation in a module).

1. Update your workspace before starting work:

   .. code-block:: shell

      west update

2. Create development branches in each module you need to modify:

   .. code-block:: shell

      cd zephyr
      git checkout -b new-api-feature

      cd ../modules/lib/zcbor
      git checkout -b support-new-api

3. Make changes across modules.

4. If you need to update other (unmodified) projects later, make sure all
   changes are committed first. Return to the workspace root before checking
   project status:

   .. code-block:: shell

      cd ../../..
      git -C zephyr status
      git -C modules/lib/zcbor status

      west update
      git -C modules/lib/zcbor checkout support-new-api

>       git -C modules/lib/zcbor rebase manifest-rev

Reviewer: Instead of doing this all manually, I suggest that this section describes that the applicable manifest `west.yml` file needs to be updated. Or better, links to the relevant existing documentation.

5. Test your changes with the updated workspace.

.. _west-module-dev-troubleshooting:

Troubleshooting
***************

.. _west-module-dev-problem-detached-head:

Problem: I am in a "detached HEAD" State
========================================

After running ``west update``, you may find yourself in a "detached HEAD"
state in some project repositories. This happens when west checks out

> ``manifest-rev`` (which is a detached reference).

Reviewer: I never had this situation when working with west. How exactly does this happen? Using west, how can I reproduce this?

Author: Hi @rettichschnidi, thanks for the review. The detached HEAD state is actually the default behavior of west update,  
 and it's documented in the official west docs.                                                                       
                                                                                                                      
 According to west-update-procedure:                                                                                  
                                                                                                                      
```
4. Checks out manifest-rev in the local working copy as a detached HEAD.                                           
                                                                                                             
For safety, west update uses git checkout --detach to check out a detached HEAD at the manifest revision for each updated project, leaving behind any branches which were already checked out.                                       
```
The reasons are explained in [west-update-detached-heads](https://docs.zephyrproject.org/latest/develop/west/why.html#west-update-detached-heads):                                                             
                                                                                                                    
1. Determinism: two users running west update on the same manifest should get the exact same workspace contents.     
2. Safety: git checkout --detach is a safe operation that won't move or lose commits on existing local branches.     
                                                                                                                    
How to reproduce:                                                                                                    
                                                                                                                    
1. Make sure the project's revision in west.yml is a SHA or tag, not a branch name.                                  
2. Run west update.                                                                                                  
3. Check the project with git status — it will show HEAD detached at <revision>.                                     
                                                                                                                    
You may not have seen this if you usually run west update --rebase / --keep-descendants, or if your manifest tracks branches and your local branch already points to the same commit. But for users running a plain west update, this is the normal state they end up in.                                                                                     
                                                                                                                    
I've updated the troubleshooting section in this PR to simply reference the existing west documentation instead of duplicating the explanation.

**Solution**: Simply check out a local branch:

.. code-block:: shell

   git checkout -b my-local-branch

Or if you already have a local branch:

.. code-block:: shell

   git checkout my-existing-branch

.. _west-module-dev-problem-lost-commits:

Problem: I think I lost my commits!
===================================

If you think ``west update`` has lost your commits, don't panic. Git does not
delete commits that are referenced by branch tips. Use ``git reflog`` to
find your commits:

.. code-block:: shell

   # View the reflog to find your commit
   git reflog

   # Once you find the commit SHA, create a branch pointing to it
   git branch recover-my-work <SHA>

.. _west-module-dev-problem-conflict:

Problem: Merge Conflicts When Rebasing
======================================

If you get merge conflicts when rebasing your local branch onto ``manifest-rev``:

.. code-block:: shell

   # Start the rebase
   git rebase manifest-rev

   # If conflicts occur, resolve them, then:
   git add <resolved-files>
   git rebase --continue

   # Or, if you want to abort the rebase:

>    git rebase --abort

Reviewer: This is regular git work flow. I do not see the value added by this section.

.. _west-module-dev-best-practices:

Best Practices
**************

1. **Always work on a local branch**. Never commit directly to ``manifest-rev``
   or work on a detached HEAD for extended periods.

2. **Commit early and often**. This makes it easier to recover work if
   something goes wrong.

3. **Run ``west update`` from anywhere in the workspace**. West will find the

>    workspace root and update all projects consistently.

Reviewer: Gut feeling: What West does internally is not relevant. Would therefore change to:
```suggestion
3. **Run ``west update`` from anywhere in the workspace**. West update all projects to the version specified in the manifest file(s).
```

4. **Use ``west status`` to check project status**. This command runs
   ``git status`` in each project repository.

5. **Understand your manifest**. Knowing which revision your manifest expects
   for each project helps you understand what ``west update`` will do.

6. **Use ``west manifest --resolve``** to see the full resolved manifest
   when using :ref:`west-manifest-import`.

.. code-block:: shell

   # See the resolved manifest (all projects and their revisions)

>    west manifest --resolve

Reviewer: This says twice the same.

.. _west-module-development:

Module Development with West
############################

This page describes how to safely develop code within west-managed modules
(project repositories) while keeping the ability to use ``west update`` to update
other dependencies. This is a common workflow for developers who contribute to
Zephyr modules or maintain their own downstream forks.

.. _west-module-dev-overview:

Overview
********

When you use west to manage a multi-repository workspace, you may need to
modify code in one or more project repositories (modules) while still being able
to update the rest of the workspace with ``west update``.

> A common concern among developers is that west's ``manifest-rev`` branch will

Reviewer: I think the concern is about ``west update`, not `manifest-rev`.

```suggestion
A common concern among developers is that running ```west update`` branch will
```

      Always work on a local branch. Never work directly on ``manifest-rev``
      or on a detached HEAD unless you know what you are doing.

4. **Make your changes and commit them**:

   .. code-block:: shell

      # ... edit files ...
      git add .
      git commit -s

5. **Update other projects in the workspace** (if needed):

   You can safely run ``west update`` while on your local branch. West will
   update the ``manifest-rev`` branch in this and other projects, but will
   **not** modify your local branch.

   .. code-block:: shell

      west update

   After running ``west update``, you may want to rebase your local branch
   on the new ``manifest-rev`` to pick up any upstream changes:

   .. code-block:: shell

      git rebase manifest-rev

6. **Push your changes** (when ready):

   .. code-block:: shell

      git push my-fork my-feature

.. _west-module-dev-before-update:

Before Running ``west update``
==============================

Before running ``west update``, it is good practice to:

* **Commit or stash all local changes**. While ``west update`` will not
  overwrite committed work on local branches, having a clean working tree
  prevents confusion.

> * **Check the current state** with ``git status`` and ``west list`` to

Reviewer: Using `west status` would be more informative because less verbose.

```suggestion
* **Check the current state** with ``git status`` and ``west status`` to
```

  understand which projects have local modifications.

Example:

.. code-block:: shell

   # Check status in the current module
   git status

   # List all projects and their revisions

>    west list

Reviewer: I think "List all projects and their revisions" is a bit misleading. It only lists what the manifest specifies. Maybe "west status is more what is wanted here?

```suggestion
   west status
```

   # Run west update
   west update

   # After update, check if you need to rebase
   git log --oneline HEAD..manifest-rev

.. _west-update-parameters:

``west update`` Parameters for Module Development
*************************************************

When developing within modules, the following ``west update`` parameters are
particularly useful.

.. _west-update-keep-descendants:

``--keep-descendants``
======================

When west updates a project, it checks out ``manifest-rev`` as a detached
HEAD by default, leaving behind any branches that were already checked out.

The ``--keep-descendants`` option tells west to keep local branches checked
out as long as they point to commits that are descendants of the new
``manifest-rev``. This is useful when you have local development branches
that already include the new manifest revision and want to keep working on
them after the update.

Example:

.. code-block:: shell

   # Update while keeping local descendant branches intact
   west update --keep-descendants

.. note::

>    ``west update --keep-descendants`` never fails because it does not try

Reviewer: This is a duplication from https://docs.zephyrproject.org/latest/develop/west/built-in.html. Would just link there.

Author: > This is a duplication from https://docs.zephyrproject.org/latest/develop/west/built-in.html. Would just link there.

OK

   to rebase: in projects where your branch diverged from the incoming
   commits, it leaves your branches behind just like a plain
   ``west update`` does. Only projects where your branch is a descendant of
   the new ``manifest-rev`` keep their checked-out branch.

.. _west-update-rebase:

``--rebase``
============

The ``--rebase`` option automatically rebases any locally checked out
branches onto the new ``manifest-rev`` after the update.

This is convenient when you have a local branch that you want to keep
up-to-date with the latest upstream revision.

Example:

.. code-block:: shell

   # Update and rebase local branches onto new manifest-rev
   west update --rebase

.. note::

   ``west update --rebase`` will fail in projects that have git conflicts
   between your branch and the new commits brought in by the manifest. You
   should resolve these conflicts as you usually do with ``git``, or use
   ``git -C <project_path> rebase --abort`` to ignore the incoming changes

>    for the moment.

Reviewer: Quite some copied stuff from https://docs.zephyrproject.org/latest/develop/west/built-in.html. Would link there.

.. _west-update-narrow:

``--narrow``
============

The ``--narrow`` option makes ``west update`` only fetch the minimum amount
of data needed to update the ``manifest-rev`` branch. This can speed up
updates when you have a slow network connection.

This is less relevant for module development specifically, but can be useful

> in large workspaces.

Reviewer: a) Would rather document this on https://docs.zephyrproject.org/latest/develop/west/built-in.html
b) Not even sure if it makes sense to mention it here.

.. _west-module-dev-scenarios:

Common Development Scenarios
****************************

.. _west-module-dev-scenario-simple-fix:

Scenario 1: Simple Bug Fix in a Module
======================================

You want to fix a bug in a Zephyr module (e.g., ``modules/lib/zcbor``)
and submit the fix upstream.

1. Navigate to the module:

   .. code-block:: shell

      cd modules/lib/zcbor

2. Create a branch for your fix:

   .. code-block:: shell

>       git checkout -b fix-uart-timeout

Reviewer: When fixing stuff upstream, better to start of from the most recent main, not some (probably outdated) `manifest-rev`.

3. Make your changes and commit:

   .. code-block:: shell

      git add .
      git commit -s

4. Run ``west update`` to make sure your workspace is up-to-date:

   .. code-block:: shell

      west update

5. Rebase on ``manifest-rev`` if needed:

   .. code-block:: shell

>       git rebase manifest-rev

Reviewer: Here and in all other scenarios: Why not running `west update` before doing any dev work, i.e. before creating a feature branch? This way, rebasing will likely not even be needed.

Problem: I think I lost my commits!
===================================

If you think ``west update`` has lost your commits, don't panic. Git does not
delete commits that are referenced by branch tips. Use ``git reflog`` to
find your commits:

.. code-block:: shell

   # View the reflog to find your commit
   git reflog

   # Once you find the commit SHA, create a branch pointing to it
   git branch recover-my-work <SHA>

.. _west-module-dev-problem-conflict:

Problem: Merge Conflicts When Rebasing
======================================

If you get merge conflicts when rebasing your local branch onto ``manifest-rev``:

.. code-block:: shell

   # Start the rebase
   git rebase manifest-rev

   # If conflicts occur, resolve them, then:
   git add <resolved-files>
   git rebase --continue

   # Or, if you want to abort the rebase:
   git rebase --abort

.. _west-module-dev-best-practices:

Best Practices
**************

1. **Always work on a local branch**. Never commit directly to ``manifest-rev``
   or work on a detached HEAD for extended periods.

2. **Commit early and often**. This makes it easier to recover work if
   something goes wrong.

> 3. **Run ``west update`` from the workspace root**. This ensures all projects

Reviewer: As long as the working directory is somewhere in the workspace it will work. No need to be at the root.

   are updated consistently.

4. **Use ``west list`` to check project status**. This command shows the

>    current revision and manifest-rev for each project.

Reviewer: `west list` shows the manifest, not the status of the module.

Reviewer: Mentioning `west status` here would be nice IMHO.

5. **Understand your manifest**. Knowing which revision your manifest expects
   for each project helps you understand what ``west update`` will do.

6. **Use ``west manifest --resolve``** to see the full resolved manifest
   when using :ref:`west-manifest-import`.

.. code-block:: shell

   # See the resolved manifest (all projects and their revisions)
   west manifest --resolve

.. _west-module-dev-summary:

Summary
*******

* West's ``manifest-rev`` branch is an internal mechanism. **Your local
  branches are safe**.
* Always create a local branch before making changes in a module.
* ``west update`` can be run safely while you are on a local branch.
* Use ``git rebase manifest-rev`` to keep your local branch up-to-date

>   with the latest upstream revision.

Reviewer: `manifest-rev` does not necessarily point to the most recent upstream version.
